src/mlpack/tests/feedforward_network_test.cpp

```diff
@@ -60,6 +60,93 @@ void TestNetwork(ModelType& model,
   BOOST_REQUIRE_LE(classificationError, classificationErrorThreshold);
 }
 
+// network1 should be allocated with `new`, and trained on some data.
+template<typename MatType = arma::mat, typename ModelType>
+void CheckCopyFunction(ModelType* network1,
```

Comment:
Please avoid using raw pointers- `std::unique_ptr` would fit here. 
And avoid deletion of pointer-parameter even more. Caller of this function may not know, that `network1` won't be valid after this call. Ownership of raw pointers is usually very unclear- don't make it worse :)

Reply:
Hi @mikkac; I appreciate that you're taking the time to review and help out here! :+1:  We use raw pointers throughout the mlpack codebase typically for the sake of speed and low-level control.  Much of the code was written before smart pointers were added to the C++ standard, actually.  There are many places throughout the codebase where we do things with pointers that don't easily fit into the smart pointer framework (at least, in some cases we have, `unique_ptr` won't do, but the reference counting of `shared_ptr` has additional overhead).

Personally I think the use of a raw pointer is fine here: the goal is to specifically deallocate all of `network1`'s memory.  You can with `unique_ptr.reset()`, too, but I don't really see a particular benefit and it would make this test code different than the virtually identical pattern that appears many places elsewhere.

I don't have a particular problem if we want to use `unique_ptr<>` more throughout the codebase, but it's a tradeoff: you get some notions of safety but it comes at the cost of readability (`ModelType*` is shorter and much more readable than `unique_ptr<ModelType>`).  So it perhaps would be something that's worth community-wide discussion.

However you are right about possible confusing semantics here, and I do think it would be useful to indicate them with a comment for the function: `network1` will be deleted.  @himanshupathak21061998 I think it could be helpful to add that if you don't mind. :+1:

Reply:
Yep, I agree that in this case, raw pointer indeed is ok- my mistake. Comment with warning on pointer deletion is fine :)
I don't know the codebase well yet, so some coding rules are unknown to me, sorry

Reply:
No worries!  We can always make changes later, but my big thing is just that we're consistent throughout the codebase. :)